zerver/views/registration.py

```diff
@@ -132,14 +132,33 @@ def accounts_register(request: HttpRequest) -> HttpResponse:
                 initial={'full_name': hesiod_name if "@" not in hesiod_name else ""},
                 realm_creation=realm_creation)
             name_validated = True
-        elif settings.POPULATE_PROFILE_VIA_LDAP:
+        elif prereg_user.full_name:
+            if prereg_user.full_name_validated:
+                request.session['authenticated_full_name'] = prereg_user.full_name
+                name_validated = True
+                form = RegistrationForm({'full_name': prereg_user.full_name},
+                                        realm_creation=realm_creation)
+            else:
+                form = RegistrationForm(initial={'full_name': prereg_user.full_name},
+                                        realm_creation=realm_creation)
+        elif 'full_name' in request.POST:
+            form = RegistrationForm(
+                initial={'full_name': request.POST.get('full_name')},
+                realm_creation=realm_creation
+            )
+        else:
+            form = RegistrationForm(realm_creation=realm_creation)
+
+        if settings.POPULATE_PROFILE_VIA_LDAP:
+            # If the user can be found in LDAP, we'll use a form with pre-filled
+            # data taken from the directory. Otherwise, we'll use the form that was
+            # create above.
```

Comment:
I'm not sure I like this structural change; before, I believe we had the invariant that at most one `RegistrationForm` was initialized in each call of this code path, and now that won't be true.  

But I suppose this change does fix incorrect behavior in a somewhat unlikely situation.

Reply:
So, here's a theory -- what if the `settings.POPULATE_PROFILE_VIA_LDAP` check happens first, but rather than going all the way through creating a `RegistrationForm`, it just sets `ldap_user` and `ldap_backend` if it finds a match, and then we preserve the original if/else sequence structure for making the `RegistrationForm`?  (Basically splitting the block below this into 2 blocks)

Reply:
changed to this model in the first commit.

I think the full name fetched from ldap should take precedence over whatever other means of delivering a suggested name to pre-fill, so I put that case as the first in the big block of conditionals that create the pre-filled form.